File: polyglot/Prototype/prototype.py
# Prototype design pattern
# Imagine that in order to instantiate certain objects we need to carry out
# expensive database queries. In order to improve performance, it would 
# be beneficial to cache whatever object is created from the database
# and return a copy of the cached object whenever a subsequent request
# for object instantiation arises.

class Cloneable:
    def clone(self):
        raise Exception("Cloneable::clone is not implemented")

class Shape(Cloneable):
    def __init__(self):
        self.ID = None
    def draw(self):
        raise Exception("Shape::draw is not implemented")
    def clone(self):
        raise Exception("Shape::clone is not implemented")

class Rectangle(Shape):
    def draw(self):
        print("Inside Rectangle::draw() method.")
    def clone(self):
        clone = Rectangle()
        clone.ID = self.ID
        return clone

class Square(Shape):
    def draw(self):
        print("Inside Square::draw() method.")
    def clone(self):
        clone = Square()
        clone.ID = self.ID
        return clone

class Circle(Shape):
    def draw(self):
        print("Inside Circle::draw() method.")
    def clone(self):
        clone = Circle()
        clone.ID = self.ID
        return clone

class PrototypeManager:
    def __init__(self):
        self.shapeMap = {}   #empty dictionary
    def getShape(self,shapeId):
        cachedShape = self.shapeMap[shapeId]
        return cachedShape.clone()
    def loadCache(self):
        rect = Rectangle()
        rect.ID = "1"
        self.shapeMap[rect.ID] = rect

        square = Square()
        square.ID = "2"
        self.shapeMap[square.ID] = square

        circle = Circle()
        circle.ID = "3"
        self.shapeMap[circle.ID] = circle

# need a prototype manager
manager = PrototypeManager()
# prototype manager registers a few prototypes
manager.loadCache()
# prototype manager can now be used as factory object
clonedShape1 = manager.getShape("1")
clonedShape1.draw()

clonedShape2 = manager.getShape("2")
clonedShape2.draw()

clonedShape3 = manager.getShape("3")
clonedShape3.draw()

File: polyglot/Prototype/Prototype.scala
// Prototype design pattern
// Imagine that in order to instantiate certain objects we need to carry out
// expensive database queries. In order to improve performance, it would 
// be beneficial to cache whatever object is created from the database
// and return a copy of the cached object whenever a subsequent request
// for object instantiation arises.

import scala.collection.immutable._ // HashMap[K,V]

abstract class Shape extends Cloneable {
  var id: String = null
  def draw()
  def getId(): String = id
  def setId(id:String) = {this.id = id}
  
  override def clone(): Object = { super.clone() }
}

class Rectangle extends Shape {
  def draw() = println("Inside Rectangle::draw() method.")
}

class Square extends Shape {
  def draw() = println("Inside Square::draw() method.")
}

class Circle extends Shape {
  def draw() = println("Inside Circle::draw() method.")
}

class PrototypeManager {
  var shapeMap = new HashMap[String, Shape]
  def getShape(shapeId: String): Shape = {
    val Some(cachedShape) = shapeMap.get(shapeId)
    return cachedShape.clone().asInstanceOf[Shape]
  }
  def loadCache() = {
    val rect = new Rectangle
    rect.setId("1")
    shapeMap += ("1" -> rect)
    
    val square = new Square
    square.setId("2")
    shapeMap += ("2" -> square)

    val circle = new Circle
    circle.setId("3")
    shapeMap += ("3" -> circle)
  } 
}

object Prototype {
  def main(args: Array[String]) = {
    // new a prototype manager
    val manager = new PrototypeManager
    // prototype manager registers a few prototypes
    manager.loadCache()
    // prototype manager can now be used as factory object
    val clonedShape1 = manager.getShape("1")
    clonedShape1.draw()
    
    val clonedShape2 = manager.getShape("2")
    clonedShape2.draw()

    val clonedShape3 = manager.getShape("3")
    clonedShape3.draw()
  }
}
